examples/step-55/step-55.cc

```diff
@@ -0,0 +1,1181 @@
+                  {
+                    cell_matrix(i,j) += (symgrad_phi_u[i]
+                                         * symgrad_phi_u[j])
+                                        * fe_values.JxW(q);
+                  }
+              }
+          }
+
+
+        for (unsigned int i=0; i<dofs_per_cell; ++i)
+          for (unsigned int j=i+1; j<dofs_per_cell; ++j)
+            cell_matrix(i,j) = cell_matrix(j,i);
+
+        cell->get_mg_dof_indices (local_dof_indices);
+
+        boundary_constraints[cell->level()]
+        .distribute_local_to_global (cell_matrix,
+                                     local_dof_indices,
+                                     mg_matrices[cell->level()]);
+
+        for (unsigned int i=0; i<dofs_per_cell; ++i)
+          for (unsigned int j=0; j<dofs_per_cell; ++j)
+            if (
+              !mg_constrained_dofs.at_refinement_edge(cell->level(),
+                                                      local_dof_indices[i])
+              || mg_constrained_dofs.at_refinement_edge(cell->level(),
+                                                        local_dof_indices[j])
+            )
+              cell_matrix(i,j) = 0;
+
+        boundary_interface_constraints[cell->level()]
+        .distribute_local_to_global (cell_matrix,
+                                     local_dof_indices,
+                                     mg_interface_matrices[cell->level()]);
+      }
+  }
+
+// @sect4{StokesProblem::solve}
+
+// This function sets up things differently based on if you want to use ILU or GMG as a preconditioner.  Both methods share
+// the same solver (GMRES) but require a different preconditioner to be assembled.  Here we time not only the entire solve
+// function, but we separately time the set-up of the preconditioner as well as the GMRES solve.
+  template <int dim>
+  void StokesProblem<dim>::solve ()
+  {
+    TimerOutput::Scope solve(computing_timer, "Solve");
+    constraints.set_zero(solution);
+
+    if (solver_type == SolverType::UMFPACK)
+      {
+        computing_timer.enter_subsection ("(UMFPACK specific)");
+        computing_timer.enter_subsection ("Solve - Initialize");
+
+        SparseDirectUMFPACK A_direct;
+        A_direct.initialize(system_matrix);
+
+        computing_timer.leave_subsection ();
+        computing_timer.leave_subsection ();
+
+        {
+          TimerOutput::Scope solve_backslash(computing_timer, "Solve - Backslash");
+          A_direct.vmult(solution, system_rhs);
+        }
+
+        constraints.distribute (solution);
+        return;
+      }
+
+    // Here we must make sure to solve for the residual with "good enough" accuracy
+    SolverControl solver_control (system_matrix.m(),
+                                  1e-10*system_rhs.l2_norm());
+    unsigned int n_iterations_A;
+    unsigned int n_iterations_S;
+
+    // This is used to pass whether or not we want to solve for A inside
+    // the preconditioner
+    const bool use_expensive = true;
```

Comment:
Did you ever try out whether or not it is faster or slower if you set this to `false`? If it is universally a win one way or the other, then it may be simpler to just remove the entire other option from the program.